push_dialog.py

```diff
@@ -87,35 +90,76 @@ def update_progress(self, sent, total):
         progress = float(sent) / total * 100
         self.progress_bar.setValue(progress)
 
-    def push_project(self):
+    def get_layer_ids_to_offline_convert(self, remote_layers, remote_save_mode):
+        layer_ids = []
+        if remote_save_mode == RemoteOptionsDialog.OFFLINE: 
+            for layer in remote_layers:
+                layer_ids.append(layer.id())
+
+        for layer in self.project_get_always_offline_layers():
+            layer_ids.append(layer.id())
+        return layer_ids
+
+
+    def push_project(self, remote_layers=None, remote_save_mode=None):
+
+        can_only_be_online_layers = self.project_get_always_online_layers()
+        if can_only_be_online_layers:
+            self.show_warning_about_layers_that_cant_work_offline(can_only_be_online_layers)
+
+        layer_ids = self.get_layer_ids_to_offline_convert(remote_layers, remote_save_mode)
+        # TODO: offline convert
+        # with rasters too?
+
+        if remote_save_mode == RemoteOptionsDialog.HYBRID:
+            self.set_hybrid_flag()
+        # TODO: show actual, more informative dialog with button
+        QtGui.QDesktopServices.openUrl(QtCore.QUrl.fromLocalFile(BASE_SAVE_LOCATION))
+
+        # this here doesn't do anything for now
         device_index = self.devices_cbx.currentIndex()
         device = self.devices[device_index][1]
         mtp = connect_device(device)
         dest = 'testFILE.qgs'
         push_file(mtp, self.project.fileName(), dest, self.update_progress)
         disconnect_device(mtp)
 
+    def show_warning_about_layers_that_cant_work_offline(self, layers):
+        layers_list = ','.join([ layer.name() for layer in layers])
+        QtGui.QMessageBox.information(self.iface.mainWindow(), 'Warning','Layers {} require a real time connection'.format(layers_list))
+
+    def set_hybrid_flag(self):
+        #TODO
+        pass
+
     def on_reload_devices_btn_clicked(self):
         self.refresh_devices()
 
     @staticmethod
-    def project_has_layers_of_given_types(types):
+    def project_get_layers_of_given_types(types):
         # can see all types via
         # QgsProviderRegistry.instance().providerList()
         map_layers = QgsMapLayerRegistry.instance().mapLayers(
         )
+        result = []
         for name, layer in map_layers.items():
             if layer.providerType() in types:
-                return True
-        return False
+                result.append(layer)
+        return result
+
+    @staticmethod
+    def project_get_always_online_layers():
+        """ Layers that can't be made offline by the offline plugin """
+        online_types = ["WFS", "wcs", "wms", "mssql", "ows"]
```

Comment:
FYI

QGIS no longer crashes when offlining raster layers (as of [right now](
https://github.com/qgis/QGIS/commit/c59350a170507cdbf563e44a28508cce3feeb2e9)). But it will just leave them untouched in the project (so no offline copy).